README.md
<img align="right" src="https://raw.github.com/cliffano/nestor-lifx/master/avatar.jpg" alt="Avatar"/>

[![Build Status](https://img.shields.io/travis/cliffano/nestor-lifx.svg)](http://travis-ci.org/cliffano/nestor-lifx)
[![Dependencies Status](https://img.shields.io/david/cliffano/nestor-lifx.svg)](http://david-dm.org/cliffano/nestor-lifx)
[![Coverage Status](https://img.shields.io/coveralls/cliffano/nestor-lifx.svg)](https://coveralls.io/r/cliffano/nestor-lifx?branch=master)
[![Published Version](https://img.shields.io/npm/v/nestor-lifx.svg)](http://www.npmjs.com/package/nestor-lifx)
<br/>
[![npm Badge](https://nodei.co/npm/nestor-lifx.png)](http://npmjs.org/package/nestor-lifx)

Nestor LIFX
-----------

Nestor LIFX is CLI for Jenkins LIFX notifier.

This is handy for monitoring Jenkins build status on a [LIFX](http://www.lifx.com) bulb.

Installation
------------

    npm install -g nestor-lifx

Usage
-----

Monitor build status and notify LIFX bulb:

    nestor-lifx run

Monitor build status of a job:

    nestor-lifx run --job <job>

Monitor build status of a view:

    nestor-lifx run --view <view>

To customise status-hue map:

    nestor-lifx run --map FAIL=0x2f20,OK=0x53f0,WARN=0xffff

Colophon
--------

[Developer's Guide](http://cliffano.github.io/developers_guide.html#nodejs)

Build reports:

* [Code complexity report](http://cliffano.github.io/nestor-lifx/complexity/plato/index.html)
* [Unit tests report](http://cliffano.github.io/nestor-lifx/test/buster.out)
* [Test coverage report](http://cliffano.github.io/nestor-lifx/coverage/buster-istanbul/lcov-report/lib/index.html)
* [Integration tests report](http://cliffano.github.io/nestor-lifx/test-integration/cmdt.out)
* [API Documentation](http://cliffano.github.io/nestor-lifx/doc/dox-foundation/index.html)

Related Projects:

* [nestor](http://github.com/cliffano/nestor) - Jenkins CLI and node.js client
* [nestor-buildlight](http://github.com/cliffano/nestor-buildlight) - CLI for Jenkins build light notifier
* [nestor-ninjablocks](http://github.com/cliffano/nestor-ninjablocks) - CLI for Jenkins Ninja Blocks notifier


CHANGELOG.md
### 0.0.3
*

### 0.0.2-pre
* 

### 0.0.1
* Initial version


test/cli.js
var bag        = require('bagofcli');
var buster     = require('buster-node');
var cli        = require('../lib/cli');
var Jenkins    = require('nestor');
var lifx       = require('lifx');
var NestorLIFX = require('../lib/nestorlifx');
var referee    = require('referee');
var assert     = referee.assert;

buster.testCase('cli - exec', {
  'should contain commands with actions': function (done) {
    var mockCommand = function (base, actions) {
      assert.defined(base);
      assert.defined(actions.commands.run.action);
      done();
    };
    this.mock({});
    this.stub(bag, 'command', mockCommand);
    cli.exec();
  }
});

buster.testCase('cli - run', {
  setUp: function () {
    this.mockConsole = this.mock(console);
    this.mockLifx    = this.mock(lifx);
    this.mockProcess = this.mock(process);

    this.mockLifx.expects('init').once().withExactArgs().returns({});
  },
  'should notify LIFX when there is no monitoring error': function (done) {
    this.stub(bag, 'command', function (base, actions) {
      actions.commands.run.action({ job: 'somejob', schedule: '* * * * * *', scheme: 'red,green,blue', map: 'fail=red,ok=green,warn=blue', usbled: '/some/usbled/path', blinkOnfailure: true });
    });
    this.stub(Jenkins.prototype, 'monitor', function (opts, cb) {
      assert.equals(opts.job, 'somejob');
      assert.equals(opts.view, undefined);
      assert.equals(opts.schedule, '* * * * * *');
      cb(null, 'ok');
    });
    this.stub(NestorLIFX.prototype, 'notify', function (result) {
      assert.equals(result, 'ok');
      done();
    });
    cli.exec();
  },
  'should monitor using default settings': function (done) {
    this.stub(bag, 'command', function (base, actions) {
      actions.commands.run.action({});
    });
    this.stub(Jenkins.prototype, 'monitor', function (opts, cb) {
      assert.equals(opts.job, undefined);
      assert.equals(opts.view, undefined);
      assert.equals(opts.schedule, undefined);
      cb(null, 'ok');
    });
    this.stub(NestorLIFX.prototype, 'notify', function (result) {
      assert.equals(result, 'ok');
      done();
    });
    cli.exec();
  },
  'should log error message and exit with non-zero code': function () {
    this.mockConsole.expects('error').once().withExactArgs('some error');
    this.mockProcess.expects('exit').once().withExactArgs(1);
    this.stub(bag, 'command', function (base, actions) {
      actions.commands.run.action({});
    });
    this.stub(Jenkins.prototype, 'monitor', function (opts, cb) {
      assert.equals(opts.job, undefined);
      assert.equals(opts.view, undefined);
      assert.equals(opts.schedule, undefined);
      cb(new Error('some error'));
    });
    cli.exec();
  }
});


test/nestorlifx.js
var buster     = require('buster-node');
var fs         = require('fs');
var lifx       = require('lifx');
var NestorLIFX = require('../lib/nestorlifx');
var referee    = require('referee');
var assert     = referee.assert;

buster.testCase('nestorlifx - nestorlifx', {
  setUp: function () {
    this.mockLifx = this.mock(lifx);
  },
  'should set opts to default when there is no customisation': function () {
    this.mockLifx.expects('init').once().withExactArgs().returns({});
    var nestor = new NestorLIFX({});
    assert.equals(nestor.opts.map, { ok: 0x53f0, fail: 0xffff, warn: 0x2f20 });
    assert.defined(nestor.lifx);
  },
  'should override status hue map if provided in opts': function () {
    this.mockLifx.expects('init').once().withExactArgs().returns({});
    var nestor = new NestorLIFX({ map: { ok: 0x1, fail: 0x2, warn: 0x3 } });
    assert.equals(nestor.opts.map, { ok: 0x1, fail: 0x2, warn: 0x3 });
    assert.defined(nestor.lifx);
  }
});

buster.testCase('nestorlifx - notify', {
  setUp: function () {
    this.mockConsole = this.mock(console);
    this.mockLifx = this.mock(lifx);
  },
  'should use mapped hue for known status': function (done) {
    var mockLifx = {
      lightsOn: function () { return; },
      lightsColour: function (hue, saturation, luminance, whiteColour, fadeTime) {
        assert.equals(hue, 0x53f0);
        assert.equals(saturation, 0xffff);
        assert.equals(luminance, 0x1000);
        assert.equals(whiteColour, 0);
        assert.equals(fadeTime, 0);
        done();
      }
    };
    this.mockLifx.expects('init').once().withExactArgs().returns(mockLifx);
    this.mockConsole.expects('log').once().withExactArgs('Setting LIFX bulb hue to %s for status %s', 0x53f0, 'ok');
    var nestor = new NestorLIFX({});
    nestor.notify('ok');
  },
  'should use predefined hue for unknown status': function (done) {
    var mockLifx = {
      lightsOn: function () { return; },
      lightsColour: function (hue, saturation, luminance, whiteColour, fadeTime) {
        assert.equals(hue, 0xaaaa);
        assert.equals(saturation, 0xffff);
        assert.equals(luminance, 0x1000);
        assert.equals(whiteColour, 0);
        assert.equals(fadeTime, 0);
        done();
      }
    };
    this.mockLifx.expects('init').once().withExactArgs().returns(mockLifx);
    this.mockConsole.expects('log').once().withExactArgs('Setting LIFX bulb hue to %s for status %s', 0xaaaa, 'inexistingstatus');
    var nestor = new NestorLIFX({});
    nestor.notify('inexistingstatus');
  }
});


lib/cli.js
var bag         = require('bagofcli');
var Jenkins     = require('nestor');
var NestorLIFX  = require('./nestorlifx');
var querystring = require('querystring');

function _run(args) {

  var lifxOpts = {
    map: args.map ? querystring.parse(args.map, ',', '=') : undefined
  };

  var jenkinsOpts = {
    job     : args.job,
    view    : args.view,
    schedule: args.schedule
  };

  var nestorLifx = new NestorLIFX(lifxOpts);
  var jenkins    = new Jenkins(process.env.JENKINS_URL);

  jenkins.monitor(jenkinsOpts, function (err, result) {
    if (err) {
      console.error(err.message);
      process.exit(1);
    } else {
      nestorLifx.notify(result);
    }
  });
}

/**
 * Execute Nestor LIFX CLI.
 */
function exec() {

  var actions = {
    commands: {
      run: { action: _run }
    }
  };

  bag.command(__dirname, actions);
}

exports.exec = exec;


complexity/plato/report.js
__report = {"summary":{"total":{"jshint":0,"sloc":87,"maintainability":134.12479124818958},"average":{"sloc":43,"maintainability":"67.06","jshint":"0.00"}},"reports":[{"info":{"file":"lib/cli.js","fileShort":"lib/cli.js","fileSafe":"lib_cli_js","link":"files/lib_cli_js/index.html"},"jshint":{"messages":0},"complexity":{"aggregate":{"sloc":{"logical":26,"physical":45},"cyclomatic":3,"halstead":{"operators":{"distinct":11,"total":64,"identifiers":["__stripped__"]},"operands":{"distinct":44,"total":78,"identifiers":["__stripped__"]},"length":142,"vocabulary":55,"difficulty":9.75,"volume":820.9530793205017,"effort":8004.292523374892,"bugs":0.27365102644016726,"time":444.68291796527177},"params":3,"line":1,"cyclomaticDensity":11.538461538461538,"complexity":{"cyclomatic":3,"sloc":{"logical":26,"physical":45},"halstead":{"operators":{"distinct":11,"total":64,"identifiers":["__stripped__"]},"operands":{"distinct":44,"total":78,"identifiers":["__stripped__"]},"length":142,"vocabulary":55,"difficulty":9.75,"volume":820.9530793205017,"effort":8004.292523374892,"bugs":0.27365102644016726,"time":444.68291796527177}}},"module":"lib/cli.js","maintainability":68.61436163992445}},{"info":{"file":"lib/nestorlifx.js","fileShort":"lib/nestorlifx.js","fileSafe":"lib_nestorlifx_js","link":"files/lib_nestorlifx_js/index.html"},"jshint":{"messages":0},"complexity":{"aggregate":{"sloc":{"logical":20,"physical":42},"cyclomatic":3,"halstead":{"operators":{"distinct":11,"total":56,"identifiers":["__stripped__"]},"operands":{"distinct":36,"total":72,"identifiers":["__stripped__"]},"length":128,"vocabulary":47,"difficulty":11,"volume":710.9873730147376,"effort":7820.861103162114,"bugs":0.23699579100491253,"time":434.4922835090063},"params":2,"line":1,"cyclomaticDensity":15,"complexity":{"cyclomatic":3,"sloc":{"logical":20,"physical":42},"halstead":{"operators":{"distinct":11,"total":56,"identifiers":["__stripped__"]},"operands":{"distinct":36,"total":72,"identifiers":["__stripped__"]},"length":128,"vocabulary":47,"difficulty":11,"volume":710.9873730147376,"effort":7820.861103162114,"bugs":0.23699579100491253,"time":434.4922835090063}}},"module":"lib/nestorlifx.js","maintainability":65.51042960826513}}]}

lib/nestorlifx.js
var lifx = require('lifx');

/**
 * class NestorLIFX
 *
 * @param {String} opts: optional
 * - map: status-hexcolour map, defaults to { OK: '00FF00', FAIL: 'FF0000', WARN: 'FFFF00' }
 */
function NestorLIFX(opts) {
  this.opts = opts;

  const MAP = {
    ok  : 0x53f0,
    fail: 0xffff,
    warn: 0x2f20
  };
  this.opts.map = this.opts.map || MAP;

  this.lifx = lifx.init();
}

/**
 * Notify build status as a colour on Ninja Blocks device LED.
 *
 * @param {String} status: build status
 */
NestorLIFX.prototype.notify = function (status) {
  const UNKNOWN      = 0xaaaa;
  const SATURATION   = 0xffff;
  const LUMINANCE    = 0x1000;
  const WHITE_COLOUR = 0;
  const FADE_TIME    = 0;

  var hue = this.opts.map[status] !== undefined ? this.opts.map[status] : UNKNOWN;

  console.log('Setting LIFX bulb hue to %s for status %s', hue, status);
  
  this.lifx.lightsOn();
  this.lifx.lightsColour(hue, SATURATION, LUMINANCE, WHITE_COLOUR, FADE_TIME);
};

module.exports = NestorLIFX;
